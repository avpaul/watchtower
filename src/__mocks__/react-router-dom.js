import React from 'react';
import PropTypes from 'prop-types';

const rrd = require('react-router-dom');

// Just render plain div with its children
rrd.BrowserRouter = ({ children }) => <div>{children}</div>;

rrd.BrowserRouter.propTypes = {
  children: PropTypes.objectOf(PropTypes.element, PropTypes.func).isRequired,
};

module.exports = rrd;
