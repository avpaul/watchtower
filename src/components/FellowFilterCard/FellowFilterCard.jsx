import React from 'react';
import PropTypes from 'prop-types';
import check from '../../static/check-mark.svg';
import defaultPicture from '../../static/profile.svg';

/**
 * @method FellowFilterCard
 * @summary - Method to render fellow's cards that can be filtered
 * @param { cardName {string},
 * numberOfFellows { number }, isTicked {object - Project|status},
 * handleClick { function }, filterKey {string} } param0 {Object}
 * @returns A rendered version of a clickable fellows card with the name, number of fellows,
 * whether the card should be ticked and a key to be filtered by.
 */
const FellowFilterCard = ({
  cardName,
  numberOfFellows,
  isTicked,
  handleCardClick,
  filterKey,
  isManager,
  picture = '',
  displayPicture
}) => {
  const shouldDisplayTick =
    isTicked.project === cardName ||
    isTicked.status === cardName ||
    isTicked.type === cardName ||
    isTicked.criteria === cardName ||
    isTicked.level === cardName ||
    isTicked.manager_email === cardName;

  const shouldRenderImage =
    isManager && displayPicture ? (
      <img className="manager-picture" src={picture || defaultPicture} alt="" />
    ) : (
      ''
    );

  const adjustCardNameStyles = isManager && displayPicture ? `pl-3` : '';
  const cardNameStyle = isManager
    ? 'manager-filter-card'
    : 'fellow-filter-card';

  const styleCardNumber = `float-right right ${
    isManager ? 'manager-card__top mb-2' : 'mt-3'
  }  card-number-display`;

  return (
    <div
      className={`fellow-summary-card ${cardNameStyle} mt-4
    ${shouldDisplayTick ? 'fellow-card-highlight' : ''}`}
      id={cardName}
      filterKey={filterKey}
      onClick={handleCardClick}
      onKeyPress={handleCardClick}
      role="button"
      tabIndex="0"
    >
      <div className="">
        <img
          className="img-responsive rounded-circle checkmark float-right"
          style={{
            display: shouldDisplayTick ? 'block' : 'none'
          }}
          src={check}
          alt=""
        />
      </div>
      <div className="float-left left-card mt-3 card-name">
        <div className="row">
          <div className="">{shouldRenderImage}</div>
          <div className="">
            <span className={adjustCardNameStyles}>{cardName}</span>
          </div>
        </div>
      </div>
      <div className={styleCardNumber}>{numberOfFellows}</div>
    </div>
  );
};

/**
 * @name FellowFilterCard Proptypes
 * @type (Proptypes)
 * @property cardName - String
 * @property numberOfFellow - number
 * @property handleCardClick - function
 * @property filterKey - string
 */
FellowFilterCard.propTypes = {
  cardName: PropTypes.string.isRequired,
  numberOfFellows: PropTypes.number.isRequired,
  isTicked: PropTypes.shape({}).isRequired,
  handleCardClick: PropTypes.func.isRequired,
  filterKey: PropTypes.string.isRequired,
  isManager: PropTypes.bool,
  picture: PropTypes.string,
  displayPicture: PropTypes.bool
};

FellowFilterCard.defaultProps = {
  picture: '',
  isManager: false,
  displayPicture: false
};

export default FellowFilterCard;
