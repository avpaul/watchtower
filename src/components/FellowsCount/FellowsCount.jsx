import React from 'react';
import PropTypes from 'prop-types';
import FilterButton from '../Buttons/Button';

const FellowsCount = ({ count, clearFilters, countName }) => {
  const resultTerm = count > 1 ? countName || 'Fellows' : countName || 'Fellow';
  return (
    <div className="ops-dashboard__fellows-summary fellow-summary-page-header">
      <span className="header-underline mr-2 pb-2">{`${count ||
        0} Total`}</span>
      <span className="mr-3">{`${resultTerm} (Filtered)`}</span>
      <FilterButton clearFilters={clearFilters} />
    </div>
  );
};

FellowsCount.propTypes = {
  count: PropTypes.number.isRequired,
  clearFilters: PropTypes.func,
  countName: PropTypes.string
};

FellowsCount.defaultProps = {
  clearFilters: () => {},
  countName: ''
};

export default FellowsCount;
