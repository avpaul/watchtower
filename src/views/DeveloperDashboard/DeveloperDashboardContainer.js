import { connect } from 'react-redux';

import ConnectedDeveloperDashboard from './DeveloperDashboard';
import getManagerFellowsSummary from '../../redux/actionCreators/managerFellowsSummaryActions';
import deactivatePIPAction from '../../redux/actionCreators/pipDeactivationActions';

export const mapStateToProps = ({ managerFellowsSummary }) => ({
  loading: managerFellowsSummary.loading,
  data: managerFellowsSummary.data,
  error: managerFellowsSummary.error
});

export default connect(
  mapStateToProps,

  { getManagerFellowsSummary, deactivatePIPAction }
)(ConnectedDeveloperDashboard);
