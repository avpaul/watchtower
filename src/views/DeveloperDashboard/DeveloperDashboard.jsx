import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Route, Switch } from 'react-router-dom';

import MapFellowsSummaryCard from '../../components/MapFellowsSummaryCard';
import Error from '../../components/Error';
import FilterButton from '../../components/Buttons/Button';
import FellowHistoryContainer from '../../components/FellowHistory';

class DeveloperDashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      fellowSummaryDetails: []
    };
    this.handleCardClick = this.handleCardClick.bind(this);
  }

  componentDidMount() {
    const { getManagerFellowsSummary, user } = this.props;

    getManagerFellowsSummary(user.roles, user.email).then(data => {
      if (!data.error) {
        switch (true) {
          case !!user.roles.WATCH_TOWER_EM:
            this.updateState(
              data.managerFellowsSummary.engineeringManager.ttls
            );
            break;
          case !!user.roles.WATCH_TOWER_SL:
            this.updateState(data.managerFellowsSummary.simulationsLead.lfs);
            break;
          case !!user.roles.WATCH_TOWER_LF:
          case !!user.roles.WATCH_TOWER_TTL:
            this.setState({
              fellowSummaryDetails: data.managerFellowsSummary.data
            });
            break;
          default:
        }
      }
    });
  }

  updateState = data => {
    const fellows = [];
    data.forEach(x => fellows.push(...x.fellows));
    this.setState({ fellowSummaryDetails: fellows });
  };

  redirectUrl = (email, history) => {
    const name = email.substr(0, email.search('@andela.com'));
    history.push(`/dashboard/fellows/${name}`);
  };

  handleCardClick = e => {
    const id = e.target.getAttribute('data-key');
    const { history } = this.props;
    const { fellowSummaryDetails } = this.state;
    const { email } = fellowSummaryDetails[id].user || fellowSummaryDetails[id];
    this.redirectUrl(email, history);
  };

  renderResultCount = () => {
    const { fellowSummaryDetails } = this.state;
    const results = fellowSummaryDetails.length || 0;
    const resultTerm = results > 1 ? 'Fellows' : 'Fellow';
    return (
      <div className="ops-dashboard__fellows-summary fellow-summary-page-header">
        <span className="header-underline mr-2 pb-2">{`${results ||
          0} Total`}</span>
        <span className="mr-3">{`${resultTerm} (Filtered)`}</span>
        <FilterButton />
      </div>
    );
  };

  renderFellowsDashboard = fellowSummaryDetails => (
    <Fragment>
      <div>{this.renderResultCount()}</div>
      <div className="">
        <MapFellowsSummaryCard
          handleClick={this.handleCardClick}
          fellowsSummaryCardDetails={fellowSummaryDetails}
        />
      </div>
    </Fragment>
  );

  render() {
    const { fellowSummaryDetails } = this.state;
    const { role, user } = this.props;

    const { ErrorBoundary } = Error;

    return (
      <ErrorBoundary>
        <Switch>
          <Route
            path="/dashboard/fellows/:name"
            render={newProps => (
              <FellowHistoryContainer
                role={role}
                user={user}
                {...newProps}
                fellowSummaryDetails={fellowSummaryDetails}
              />
            )}
          />

          <Route
            path="/dashboard/fellows"
            component={() => this.renderFellowsDashboard(fellowSummaryDetails)}
          />
        </Switch>
      </ErrorBoundary>
    );
  }
}

DeveloperDashboard.propTypes = {
  getManagerFellowsSummary: PropTypes.func.isRequired,
  user: PropTypes.shape({}).isRequired,
  role: PropTypes.shape({}).isRequired,
  history: PropTypes.shape({}).isRequired,
  data: PropTypes.shape({}).isRequired
};

export default DeveloperDashboard;
