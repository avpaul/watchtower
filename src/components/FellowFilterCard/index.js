import FellowFilterCard from './FellowFilterCard';

export default FellowFilterCard;
