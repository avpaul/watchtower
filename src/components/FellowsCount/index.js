import FellowsCount from './FellowsCount';

export default FellowsCount;
