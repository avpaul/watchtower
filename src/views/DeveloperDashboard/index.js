import DeveloperDashboardContainer from './DeveloperDashboardContainer';

export default DeveloperDashboardContainer;
