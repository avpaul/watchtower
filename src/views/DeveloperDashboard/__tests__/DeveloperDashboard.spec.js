import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import DeveloperDashboard from '../DeveloperDashboard';
import FellowSummaryData from '../../../__mocks__/fellowSummary.json';
import SimulationLeadData from '../../../__mocks__/simulationsLeadLf.json';
import EngineeringManagerDeveloperSummary from '../../../__mocks__/emDeveloperSummery.json';
import FilterButton from '../../../components/Buttons/Button';
import FellowFilterCard from '../../../components/FellowFilterCard';
import MapFellowsFilterCard from '../../../components/MapFellowsFilterCard';
import FellowsCount from '../../../components/FellowsCount';

const getManagerDataByRole = loggedInRole => {
  switch (loggedInRole) {
    case 'WATCH_TOWER_EM':
      return EngineeringManagerDeveloperSummary;
    case 'WATCH_TOWER_SL':
      return SimulationLeadData;
    case 'WATCH_TOWER_TTL':
    case 'WATCH_TOWER_LF':
      return { data: FellowSummaryData };
    default:
      return null;
  }
};

/**
 ** Creates an enzyme instance to test the component.
 * @function
 * @param loggedInRole The role of the logged in user
 * @param urlPath The test url used to render the components accordingly.
 * @return { developerDashboardWrapper, props }
 */
const setup = (loggedInRole, urlPath = '/developers') => {
  const managerDataForTest = getManagerDataByRole(loggedInRole);

  const props = {
    history: { push: jest.fn() },
    user: {
      roles: { [`${loggedInRole}`]: '34323234Yf-34' },
      email: 'ty@andela.com'
    },
    role: loggedInRole,
    getManagerFellowsSummary: jest.fn().mockImplementation(() =>
      Promise.resolve({
        error: false,
        managerFellowsSummary: managerDataForTest
      })
    )
  };

  const developerDashboardWrapper = shallow(<DeveloperDashboard {...props} />);

  const developerDashboardMountWrapper = shallow(
    <MemoryRouter initialEntries={[urlPath]}>
      {developerDashboardWrapper
        .find('Route[path="/developers"]')
        .prop('render')()}
    </MemoryRouter>
  );
  developerDashboardWrapper.setState({ fellowSummaryDetails: [] });
  return { developerDashboardWrapper, props, developerDashboardMountWrapper };
};

describe('Developers dashboard test', () => {
  /**
   ** Tests the Developer Dashboard according to the provided user role.
   * @param role The user role
   */
  const testComponentUsingRole = role => {
    const { developerDashboardWrapper } = setup(role);
    expect(developerDashboardWrapper).toMatchSnapshot();
  };

  const testCardClick = (currentTarget, isTickedResult, fellowLength) => {
    const { developerDashboardWrapper, developerDashboardMountWrapper } = setup(
      'WATCH_TOWER_TTL'
    );
    developerDashboardWrapper.setState({
      isTicked: { status: 'On Track', project: 'All Products' },
      allFellows: FellowSummaryData,
      fellowSummaryDetails: []
    });
    developerDashboardMountWrapper
      .find(MapFellowsFilterCard)
      .at(1)
      .dive()
      .find(FellowFilterCard)
      .first()
      .dive()
      .simulate('click', {
        currentTarget: {
          id: currentTarget.id,
          attributes: [{}, {}, { value: currentTarget.value }]
        }
      });
    expect(developerDashboardWrapper.state('isTicked')).toEqual(isTickedResult);
    expect(
      developerDashboardWrapper.state('fellowSummaryDetails').length
    ).toEqual(fellowLength);
  };

  it('should render developers dashboard as WATCH_TOWER_TTL without crashing', () => {
    testComponentUsingRole('WATCH_TOWER_TTL');
  });
  it('should render developers dashboard without crashing', () => {
    const { developerDashboardWrapper } = setup('WATCH_TOWER_LF');
    expect(developerDashboardWrapper).toMatchSnapshot();
    developerDashboardWrapper.setState({
      lfTtlSummary: [{ id: 1, name: '' }],
      isTicked: { status: 'status' },
      resetFellows: [{ id: 1, status: 'status' }],
      allFellows: []
    });
    developerDashboardWrapper.instance().filterFellows('main');
    developerDashboardWrapper.instance().filterFellows(1);
    developerDashboardWrapper.instance().renderResultCount();
    developerDashboardWrapper.instance().mapLfTtlData();
    developerDashboardWrapper.instance().renderFellowsDashboard([], 1);
    developerDashboardWrapper
      .instance()
      .updateState(EngineeringManagerDeveloperSummary.managees);
    developerDashboardWrapper.instance().filterFellows('-LQnPSlfF0FLe3tayNGy');
  });

  it('should call filterFellows when the status filter represents all fellows', () => {
    const { developerDashboardWrapper } = setup('WATCH_TOWER_LF');
    developerDashboardWrapper.setState({
      isTicked: { status: 'All Fellows' },
      resetFellows: []
    });
    developerDashboardWrapper.instance().filterFellows(1);
  });

  it('should render developers dashboard without crashing', () => {
    const { developerDashboardWrapper } = setup('WATCH_TOWER_SL');
    expect(developerDashboardWrapper).toMatchSnapshot();
  });

  it('should call the getManagerFellowsAction when developers dashboard mounts', () => {
    const { developerDashboardWrapper, props } = setup('WATCH_TOWER_EM');
    const { getManagerFellowsSummary } = props;
    developerDashboardWrapper.instance().componentDidMount();
    expect(getManagerFellowsSummary).toHaveBeenCalled();
  });

  it('should redirect user to developer page when developer card is clicked', () => {
    const { props } = setup('WATCH_TOWER_LF');
    const wrapper = shallow(<DeveloperDashboard {...props} />);
    const instance = wrapper.instance();
    const fellowSummaryDetails = [
      {
        user: {
          email: 'brian.mboya@andela.com',
          fellow_id: '-LP6C8U9vaZuCUteSlXq'
        }
      }
    ];
    instance.setState({ fellowSummaryDetails });
    const event = {
      currentTarget: {
        id: 0
      }
    };

    instance.handleCardClick(event);
    instance.redirectUrl(fellowSummaryDetails[0].user.fellow_id, props.history);
    expect(props.history.push).toHaveBeenCalled();
  });

  it('should clear state to default when clear filters button is clicked', () => {
    const { developerDashboardWrapper, developerDashboardMountWrapper } = setup(
      'WATCH_TOWER_TTL'
    );
    developerDashboardWrapper.setState({
      isTicked: { status: 'On Track', project: 'All Product' },
      fellowSummaryDetails: []
    });
    developerDashboardMountWrapper
      .find(FellowsCount)
      .dive()
      .find(FilterButton)
      .dive()
      .simulate('click');
    expect(developerDashboardWrapper.state('isTicked')).toEqual({
      managers: 'All Managers',
      project: 'All Products',
      status: 'All Fellows'
    });
  });

  it('should set status to all fellows  when a all-fellows card is clicked', () => {
    testCardClick(
      {
        id: 'All Fellows',
        value: 'status'
      },
      {
        project: 'All Products',
        status: 'All Fellows'
      },
      6
    );
  });

  it('should status to ontrack when a on-track card is clicked', () => {
    testCardClick(
      {
        id: 'Watch Tower',
        value: 'project'
      },
      { status: 'On Track', project: 'Watch Tower' },
      2
    );
  });

  it('should set status to Off Track when Off Track card is clicked', () => {
    testCardClick(
      {
        id: 'Off Track',
        value: 'status'
      },
      {
        project: 'All Products',
        status: 'Off Track'
      },
      4
    );
  });

  it('should set status to PIP when pip card is clicked', () => {
    testCardClick(
      {
        id: 'PIP',
        value: 'status'
      },
      {
        project: 'All Products',
        status: 'PIP'
      },
      0
    );
  });

  it('should filter Fellows given the manager ID', () => {});
});

export default getManagerDataByRole;
