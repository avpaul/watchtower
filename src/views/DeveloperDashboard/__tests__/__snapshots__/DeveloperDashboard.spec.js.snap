// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Developers dashboard test should render developers dashboard as WATCH_TOWER_TTL without crashing 1`] = `
<ErrorBoundary>
  <Switch>
    <Route
      path="/developers/pip/activation/:id"
      render={[Function]}
    />
    <Route
      path="/developers/:id"
      render={[Function]}
    />
    <Route
      path="/developers"
      render={[Function]}
    />
  </Switch>
</ErrorBoundary>
`;

exports[`Developers dashboard test should render developers dashboard without crashing 1`] = `
<ErrorBoundary>
  <Switch>
    <Route
      path="/developers/pip/activation/:id"
      render={[Function]}
    />
    <Route
      path="/developers/:id"
      render={[Function]}
    />
    <Route
      path="/developers"
      render={[Function]}
    />
  </Switch>
</ErrorBoundary>
`;

exports[`Developers dashboard test should render developers dashboard without crashing 2`] = `
<ErrorBoundary>
  <Switch>
    <Route
      path="/developers/pip/activation/:id"
      render={[Function]}
    />
    <Route
      path="/developers/:id"
      render={[Function]}
    />
    <Route
      path="/developers"
      render={[Function]}
    />
  </Switch>
</ErrorBoundary>
`;

exports[`Developers dashboard test should render paginated developers dashboard without crashing 1`] = `
<PaginationFrontendWrapper
  component={
    <DeveloperDashboard
      getManagerFellowsSummary={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "WATCH_TOWER_LF": "34323234Yf-34",
              },
              "ty@andela.com",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Promise {},
            },
          ],
        }
      }
      history={
        Object {
          "push": [MockFunction],
        }
      }
      paginationWrapper={
        Object {
          "renderPagination": [Function],
          "state": Object {
            "paginatedData": Object {},
          },
          "updateData": [Function],
        }
      }
      role="WATCH_TOWER_LF"
      user={
        Object {
          "email": "ty@andela.com",
          "roles": Object {
            "WATCH_TOWER_LF": "34323234Yf-34",
          },
        }
      }
    />
  }
/>
`;
